### PR TITLE
relay(dm): cancel when relay meet error to close goroutine

### LINKED ISSUE #6193: gorouting & connection leak when relay meet error
### What did you do?

same relay config for different worker.

### What did you expect to see?

no gorouting/connection leak

### What did you see instead?

```
[2022/07/05 21:17:43.331 +08:00] [ERROR] [relay.go:721] ["flush meta"] [component="relay log"] [error="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta2204532292: no such file or directory"] [errorVerbose="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta2204532292: no such file or directory\ngithub.com/pingcap/ticdc/dm/pkg/terror.(*Error).Delegate\n\tgithub.com/pingcap/ticdc/dm/pkg/terror/terror.go:311\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).doFlush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:247\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).Flush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:228\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).FlushMeta\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:844\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).doIntervalOps\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:719\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[2022/07/05 21:17:43.331 +08:00] [ERROR] [relay.go:721] ["flush meta"] [component="relay log"] [error="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta2039700688: no such file or directory"] [errorVerbose="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta2039700688: no such file or directory\ngithub.com/pingcap/ticdc/dm/pkg/terror.(*Error).Delegate\n\tgithub.com/pingcap/ticdc/dm/pkg/terror/terror.go:311\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).doFlush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:247\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).Flush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:228\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).FlushMeta\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:844\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).doIntervalOps\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:719\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[2022/07/05 21:17:43.331 +08:00] [ERROR] [relay.go:721] ["flush meta"] [component="relay log"] [error="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta466240616: no such file or directory"] [errorVerbose="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta466240616: no such file or directory\ngithub.com/pingcap/ticdc/dm/pkg/terror.(*Error).Delegate\n\tgithub.com/pingcap/ticdc/dm/pkg/terror/terror.go:311\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).doFlush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:247\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).Flush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:228\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).FlushMeta\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:844\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).doIntervalOps\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:719\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[2022/07/05 21:17:43.331 +08:00] [ERROR] [relay.go:721] ["flush meta"] [component="relay log"] [error="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta1663811013: no such file or directory"] [errorVerbose="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta1663811013: no such file or directory\ngithub.com/pingcap/ticdc/dm/pkg/terror.(*Error).Delegate\n\tgithub.com/pingcap/ticdc/dm/pkg/terror/terror.go:311\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).doFlush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:247\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).Flush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:228\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).FlushMeta\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:844\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).doIntervalOps\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:719\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[2022/07/05 21:17:43.331 +08:00] [ERROR] [relay.go:721] ["flush meta"] [component="relay log"] [error="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta1749840261: no such file or directory"] [errorVerbose="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta1749840261: no such file or directory\ngithub.com/pingcap/ticdc/dm/pkg/terror.(*Error).Delegate\n\tgithub.com/pingcap/ticdc/dm/pkg/terror/terror.go:311\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).doFlush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:247\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).Flush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:228\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).FlushMeta\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:844\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).doIntervalOps\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:719\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[2022/07/05 21:17:43.331 +08:00] [ERROR] [relay.go:721] ["flush meta"] [component="relay log"] [error="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta874705000: no such file or directory"] [errorVerbose="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta874705000: no such file or directory\ngithub.com/pingcap/ticdc/dm/pkg/terror.(*Error).Delegate\n\tgithub.com/pingcap/ticdc/dm/pkg/terror/terror.go:311\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).doFlush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:247\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).Flush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:228\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).FlushMeta\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:844\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).doIntervalOps\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:719\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[2022/07/05 21:17:43.331 +08:00] [ERROR] [relay.go:721] ["flush meta"] [component="relay log"] [error="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta323006719: no such file or directory"] [errorVerbose="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta323006719: no such file or directory\ngithub.com/pingcap/ticdc/dm/pkg/terror.(*Error).Delegate\n\tgithub.com/pingcap/ticdc/dm/pkg/terror/terror.go:311\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).doFlush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:247\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).Flush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:228\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).FlushMeta\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:844\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).doIntervalOps\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:719\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[2022/07/05 21:17:43.331 +08:00] [ERROR] [relay.go:721] ["flush meta"] [component="relay log"] [error="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta1752378300: no such file or directory"] [errorVerbose="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta1752378300: no such file or directory\ngithub.com/pingcap/ticdc/dm/pkg/terror.(*Error).Delegate\n\tgithub.com/pingcap/ticdc/dm/pkg/terror/terror.go:311\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).doFlush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:247\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).Flush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:228\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).FlushMeta\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:844\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).doIntervalOps\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:719\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[2022/07/05 21:17:43.331 +08:00] [ERROR] [relay.go:721] ["flush meta"] [component="relay log"] [error="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta1377781961: no such file or directory"] [errorVerbose="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta1377781961: no such file or directory\ngithub.com/pingcap/ticdc/dm/pkg/terror.(*Error).Delegate\n\tgithub.com/pingcap/ticdc/dm/pkg/terror/terror.go:311\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).doFlush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:247\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).Flush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:228\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).FlushMeta\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:844\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).doIntervalOps\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:719\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[2022/07/05 21:17:43.331 +08:00] [ERROR] [relay.go:721] ["flush meta"] [component="relay log"] [error="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta74353850: no such file or directory"] [errorVerbose="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta74353850: no such file or directory\ngithub.com/pingcap/ticdc/dm/pkg/terror.(*Error).Delegate\n\tgithub.com/pingcap/ticdc/dm/pkg/terror/terror.go:311\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).doFlush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:247\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).Flush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:228\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).FlushMeta\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:844\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).doIntervalOps\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:719\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[2022/07/05 21:17:43.331 +08:00] [ERROR] [relay.go:721] ["flush meta"] [component="relay log"] [error="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta2873734196: no such file or directory"] [errorVerbose="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta2873734196: no such file or directory\ngithub.com/pingcap/ticdc/dm/pkg/terror.(*Error).Delegate\n\tgithub.com/pingcap/ticdc/dm/pkg/terror/terror.go:311\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).doFlush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:247\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).Flush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:228\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).FlushMeta\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:844\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).doIntervalOps\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:719\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[2022/07/05 21:17:43.331 +08:00] [ERROR] [relay.go:721] ["flush meta"] [component="relay log"] [error="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta339862555: no such file or directory"] [errorVerbose="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta339862555: no such file or directory\ngithub.com/pingcap/ticdc/dm/pkg/terror.(*Error).Delegate\n\tgithub.com/pingcap/ticdc/dm/pkg/terror/terror.go:311\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).doFlush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:247\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).Flush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:228\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).FlushMeta\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:844\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).doIntervalOps\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:719\nruntime.goexit\n\truntime/asm_amd64.s:1371"]
[2022/07/05 21:17:43.331 +08:00] [ERROR] [relay.go:721] ["flush meta"] [component="relay log"] [error="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta1841116126: no such file or directory"] [errorVerbose="[code=30006:class=relay-unit:scope=internal:level=high], Message: flush local meta, RawCause: open /tidbrlc/relay-dir/f0752560-90de-11e9-b9f6-b496914ca318.000001/relay.meta1841116126: no such file or directory\ngithub.com/pingcap/ticdc/dm/pkg/terror.(*Error).Delegate\n\tgithub.com/pingcap/ticdc/dm/pkg/terror/terror.go:311\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).doFlush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:247\ngithub.com/pingcap/ticdc/dm/relay.(*LocalMeta).Flush\n\tgithub.com/pingcap/ticdc/dm/relay/meta.go:228\ngithub.com/pingcap/ticdc/dm/relay.(*Relay).FlushMeta\n\tgithub.com/pingcap/ticdc/dm/relay/relay.go:844\ngithub.com/pingcap/ticdc/dm/relay
```
```
goroutine 873481 [select]:
[github.com/pingcap/ticdc/dm/relay.(*Relay).doIntervalOps](http://github.com/pingcap/ticdc/dm/relay.(*Relay).doIntervalOps)(0xc000cc2000, 0x3b9f908, 0xc001682d80)
	[github.com/pingcap/ticdc/dm/relay/relay.go:711](http://github.com/pingcap/ticdc/dm/relay/relay.go:711) +0x1f5
created by [github.com/pingcap/ticdc/dm/relay.(*Relay).process](http://github.com/pingcap/ticdc/dm/relay.(*Relay).process)
	[github.com/pingcap/ticdc/dm/relay/relay.go:305](http://github.com/pingcap/ticdc/dm/relay/relay.go:305) +0x57a
```

### Versions of the cluster

DM version (run `dmctl -V` or `dm-worker -V` or `dm-master -V`):
```console
(paste DM version here, and you must ensure versions of dmctl, DM-worker and DM-master are same)
```
Upstream MySQL/MariaDB server version:
```console
(paste upstream MySQL/MariaDB server version here)
```
Downstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):
```console
(paste TiDB cluster version here)
```
How did you deploy DM: tiup or manually?
```console
(leave TiUP or manually here)
```
Other interesting information (system version, hardware config, etc):
```console
>
>
```

### current status of DM cluster (execute `query-status <task-name>` in dmctl)

```console
(paste current status of DM cluster here)
```

### PR DESCRIPTION
**Pull request by lance6716**
Signed-off-by: lance6716 <lance6716@gmail.com>

### What problem does this PR solve?

Issue Number: close #6193 

### What is changed and how it works?

as title

#### Tests 

 - Integration test

### Release note 

```release-note
Fix a bug that relay goroutine and upstream connections may leak when relay meet error
```

### CODE REVIEW

**Review by D3Hunter — APPROVED** [review]

File: dm/tests/new_relay/run.sh [hunk 1]
```diff
@@ -90,6 +90,38 @@ function test_restart_relay_status() {
...
+		"RelayGetEventFailed" 1
+
+	check_log_contain_with_retry 'dispatch auto resume relay' $WORK_DIR/worker1/log/dm-worker.log
+
+	count=$(curl "http://127.0.0.1:8262/debug/pprof/goroutine?debug=2" 2>/dev/null | grep -c doIntervalOps || true)
+	if [ $count -gt 1 ]; then
+		echo "relay goroutine leak detected, count expect 1 but got $count"
+		exit 1
+	fi
+	echo ">>>>>>>>>>>>>>>>>>>>>>>>>>test test_cant_dail_upstream passed"
```
Comment (buchuitoudegou) [inline dm/tests/new_relay/run.sh:122]: test_cant_dail_upstream? Is it another test?

**Review by buchuitoudegou — APPROVED** [review]
rest lgtm

File: dm/tests/new_relay/run.sh (hunk 1 above)
Comment (lance6716) [inline dm/tests/new_relay/run.sh:122]:
```suggestion
	echo ">>>>>>>>>>>>>>>>>>>>>>>>>>test test_relay_leak passed"
```

**Pull request merged**